**Veri Yükleme**

In [2]:
import pandas as pd 
df= pd.read_csv ("C:\\Users\\nilay\\Documents\\GitHub\\Turkcell_ML_Projects-\\Advertising.csv")
df=df.drop("Unnamed: 0",axis=1)
df.head()

TV  radio  newspaper  sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

In [3]:
X=df.drop("sales",axis=1)
y=df["sales"] # numpy array
# y=df[["sales"]] # pandas dataframe

In [8]:
! pip install statsmodels


[notice] A new release of pip is available: 24.2 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [11]:
import statsmodels.api as sm 
lm=sm.OLS(y,X)
model=lm.fit()


In [ ]:
model.summary()
"""
R^2 : bağımsız değişkenlerin bağımlı değişkenleri açıklayabilme yüzdesi
Adj R-squared : Düzeltilmiş R^2 (modele değişken ekledikçe R ^2 değişkeni formülasyonundan dolayı artma eğilimindedir.Ancak bu yeni eklenen değişkenlerin modelin açıklanabilrliğine doğrudan etkisi olmayabilir)
F- statistic :modelin anlamlılığını test eder. 
Prob(F-statistic : Anlamlı bir modelin f-istatistiği 0.05 den küçük olmalıdır)
coef: b1,b2,b3 katsayılarıdır.
std err : standart error  :Katsayıların standart hatalarıdır 
t ve P>|t| : katsayıların anlamlılığını test eder. 0.05 den küçük olduğu için anlamlıdır.
[0.025 ve 0.975] : güven aralığıdır.

Tv deki 1 birimlik artışın satışlara etkisi 0.0538 birimlik artışa sebep olur.
Radyo harcamalarındaki 1 birimlik artışın etkisi 0.222 birimlik artışa sebep olur. 
"""

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                  sales   R-squared (uncentered):                   0.982
Model:                            OLS   Adj. R-squared (uncentered):              0.982
Method:                 Least Squares   F-statistic:                              3566.
Date:                Sun, 02 Mar 2025   Prob (F-statistic):                   2.43e-171
Time:                        17:38:38   Log-Likelihood:                         -423.54
No. Observations:                 200   AIC:                                      853.1
Df Residuals:                     197   BIC:                                      863.0
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
TV             0.0538      0.001     40.507      0.000       0.051       0.056
radio          0.2222      0.009     23.595      0.000       0.204       0.241
newspaper      0.0168      0.007      2.517      0.013       0.004       0.030
==============================================================================
Omnibus:                        5.982   Durbin-Watson:                   2.038
Prob(Omnibus):                  0.050   Jarque-Bera (JB):                7.039
Skew:                          -0.232   Prob(JB):                       0.0296
Kurtosis:                       3.794   Cond. No.                         12.6
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [20]:
##scikit learn ile model kurmak 
from sklearn.linear_model import LinearRegression
lr=LinearRegression()
model=lr.fit(X,y)
tahmin=model.predict(X)
df["tahmin"]=tahmin
df["hata"]=y-tahmin
df.head(20)

TV  radio  newspaper  sales     tahmin      hata
0   230.1   37.8       69.2   22.1  20.523974  1.576026
1    44.5   39.3       45.1   10.4  12.337855 -1.937855
2    17.2   45.9       69.3    9.3  12.307671 -3.007671
3   151.5   41.3       58.5   18.5  17.597830  0.902170
4   180.8   10.8       58.4   12.9  13.188672 -0.288672
5     8.7   48.9       75.0    7.2  12.478348 -5.278348
6    57.5   32.8       23.5   11.8  11.729760  0.070240
7   120.2   19.6       11.6   13.2  12.122953  1.077047
8     8.6    2.1        1.0    4.8   3.727341  1.072659
9   199.8    2.6       21.2   10.6  12.550849 -1.950849
10   66.1    5.8       24.2    8.6   7.032299  1.567701
11  214.7   24.0        4.0   17.4  17.285129  0.114871
12   23.8   35.1       65.9    9.2  10.577121 -1.377121
13   97.5    7.6        7.2    9.7   8.826300  0.873700
14  204.1   32.9       46.0   19.0  18.434366  0.565634
15  195.4   47.7       52.9   22.4  20.819300  1.580700
16   67.8   36.6      114.0   12.5  12.823657 -0.323657
17  281.4   39.6       55.8   24.4  23.224957  1.175043
18   69.2   20.5       18.3   11.3   9.951682  1.348318
19  147.3   23.9       19.1   14.6  14.166073  0.433927

In [15]:
model.intercept_

2.938889369459412

In [16]:
model.coef_

array([ 0.04576465,  0.18853002, -0.00103749])

NOT : Sklearn kütüphanesi ile eğitilen modelin katsayıları ile statsmodel ile eğitilen modelin katsayıları birbirinden farklıdır. Bunun sebebi her iki yöntemde kullanılan katsayı tahminleme metotları birbirinden farklıdır.

**Sales=2.94 + TV*0.04 + radio*0.19-newspaper*0.001**

In [23]:
df["TV"]=30
df["radio"]=40
df["newspaper"]=10
veri=[[30],[40],[10]]
veri=pd.DataFrame(veri).T
print(veri)


    0   1   2
0  30  40  10


**1 gözlem için tahmin üretme**

In [ ]:
tahmin= model.predict(veri)
print(tahmin)

[11.84265448]


c:\Users\nilay\AppData\Local\Programs\Python\Python311\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


**Tüm Gözlemler için tahmin üretme**

In [29]:
from sklearn.metrics import mean_squared_error,mean_absolute_error,r2_score
import numpy as np

tahmin2=model.predict(X)
mae=mean_absolute_error(y,tahmin2)
mse=mean_squared_error(y,tahmin2)
rmse=np.sqrt(mean_squared_error(y,tahmin2))
r2=r2_score(y,tahmin2)

print("mae:",mae)
print("mse:",mse)
print("rmse:",rmse)
print("r2:",r2)

mae: 1.2520112296870685
mse: 2.784126314510936
rmse: 1.6685701407225697
r2: 0.8972106381789522


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error,mean_squared_error,r2_score
import numpy as np

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=99) # random state verilmezse her seferinde farklı veri seti oluşur.  
model=lr.fit(X_train,y_train)
tahmin=model.predict(X_test)

#training veriseti değerlendirme
mae=mean_absolute_error(y_train,model.predict(X_train))
mse=mean_squared_error(y_train,model.predict(X_train))
rmse=np.sqrt(mean_squared_error(y_train,model.predict(X_train)))
r2=r2_score(y_train,model.predict(X_train))
print("Training veri seti değerlendirme")
print("mae:",mae)
print("mse:",mse)
print("rmse:",rmse)
print("r2:",r2)


#test veri seti değerlendirme
print("Test veri seti değerlendirme")
mae=mean_absolute_error(y_test,tahmin)
mse=mean_squared_error(y_test,tahmin)
rmse=np.sqrt(mean_squared_error(y_test,tahmin))
r2=r2_score(y_test,tahmin)

print("mae:",mae)
print("mse:",mse)
print("rmse:",rmse)
print("r2:",r2)

Training veri seti değerlendirme
mae: 1.287082714282572
mse: 2.971081299667491
rmse: 1.7236824822650751
r2: 0.8906288862925659
Test veri seti değerlendirme
mae: 1.1458629205525597
mse: 2.0485576116405175
rmse: 1.4312783138301641
r2: 0.9200846680148507


In [40]:
#k-katlı cross validasyon 
from sklearn.model_selection import cross_val_score

cv= cross_val_score(model,X_train,y_train,cv=10,scoring="r2")
print(cv)
mean=np.mean(cv)
print(mean)

cv2= cross_val_score(model,X_test,y_test,cv=10,scoring="r2")
print(cv2)
mean2=np.mean(cv2)
print(mean2)


[0.91245644 0.92453248 0.85814127 0.92005307 0.8783077  0.93204044
 0.87656032 0.5495448  0.90023107 0.79649734]
0.8548364931257494
[-3.13161617  0.94441908  0.72362936  0.64872846  0.76964744  0.92765282
  0.92591242  0.95135753  0.98675672  0.70728269]
0.44537703551296753
